edge-util/src/Microsoft.Azure.Devices.Edge.Util/Option.cs

```diff
@@ -16,6 +16,11 @@ public struct Option<T> : IEquatable<Option<T>>
 
         internal Option(T value, bool hasValue)
         {
+            if (hasValue && value == null)
+            {
+                throw new ArgumentOutOfRangeException(nameof(value), "Value cannot be null if hasValue flag is set to true.");
```

Comment:
`ArgumentNullException` ?

Reply:
Sure. Will change.

Reply:
We have a `Preconditions.CheckNotNull` which should do the right thing.

You should be able to do:
```csharp
internal Option(T value, bool hasValue)
{
    this.Value = Preconditions.CheckNotNull(value, nameof(value));
    this.hasValue = hasValue;
}
```

Reply:
Calling Precondition/checking null only if hasValue is true.  That's why I cannot use CheckNotNull.  If we want to enforce default(T) if hasValue is false. I can use CheckArgument method.